In [1]:
import pandas as pd
hollywood_movies = pd.read_csv("Hollywood_movies_dataset")
indian_movies = pd.read_csv("Indian_movies_dataset")

In [57]:
Movies_dataset = pd.concat([hollywood_movies, indian_movies], ignore_index=True)
Movies_dataset

Movie Name  Movie Release Year  \
0                                           GoodFellas                1990   
1                                           Home Alone                1990   
2                                  Edward Scissorhands                1990   
3                                         Pretty Woman                1990   
4                          Back to the Future Part III                1990   
5                                               Misery                1990   
6                               The Godfather Part III                1990   
7                                         Total Recall                1990   
8                                        Wild at Heart                1990   
9                                     Paris Is Burning                1990   
10                                            Close-Up                1990   
11                                      Jacob’s Ladder                1990   
12                                               Ghost                1990   
13                                                  It                1990   
14                                             Tremors                1990   
15                                              Dreams                1990   
16                                   Miller’s Crossing                1990   
17                                          Die Hard 2                1990   
18                            The Silence of the Lambs                1991   
19                          Terminator 2: Judgment Day                1991   
20                                Beauty and the Beast                1991   
21                                     Thelma & Louise                1991   
22                                   The Addams Family                1991   
23                                My Own Private Idaho                1991   
24                                         Point Break                1991   
25                                           Cape Fear                1991   
26                                                Hook                1991   
27                                     Boyz n the Hood                1991   
28                               A Brighter Summer Day                1991   
29                                         Barton Fink                1991   
30                        The Double Life of Véronique                1991   
31                                             My Girl                1991   
32                                      Only Yesterday                1991   
33                                      Night on Earth                1991   
34                                                 JFK                1991   
35                                         Naked Lunch                1991   
36                                      Reservoir Dogs                1992   
37                                             Aladdin                1992   
38                      Home Alone 2: Lost in New York                1992   
39                               Bram Stoker’s Dracula                1992   
40                                         Porco Rosso                1992   
41                                      Batman Returns                1992   
42                       Twin Peaks: Fire Walk with Me                1992   
43                                   Death Becomes Her                1992   
44                                              Alien³                1992   
45                                    Army of Darkness                1992   
46                                            Candyman                1992   
47                                          Unforgiven                1992   
48                                     My Cousin Vinny                1992   
49                                    Scent of a Woman                1992   
50                                      Basic Instinct            

In [3]:
Movies_dataset = Movies_dataset.sort_values(by = "Movie Release Year", ascending=True)
Movies_dataset.index = range(1,len(Movies_dataset)+1)
Movies_dataset

Movie Name  Movie Release Year   Runtime         Rating  \
1                GoodFellas                1990  145 mins  4.47 out of 5   
2                Die Hard 2                1990  124 mins  3.32 out of 5   
3         Miller’s Crossing                1990  115 mins  4.00 out of 5   
4                    Dreams                1990  119 mins  4.19 out of 5   
5                   Tremors                1990  96  mins  3.52 out of 5   
..                      ...                 ...       ...            ...   
941  A Quiet Place: Day One                2024  99  mins  3.19 out of 5   
942      Love Lies Bleeding                2024  104 mins  3.57 out of 5   
943       I Saw the TV Glow                2024  100 mins  3.53 out of 5   
944   Saripodhaa Sanivaaram                2024  166 mins            3.5   
945                Aavesham                2024  155 mins            3.9   

             Director                                        Description  \
1     Martin Scorsese  The true story of Henry Hill, a half-Irish, ha...   
2        Renny Harlin  One year after his heroics in Los Angeles, Joh...   
3           Joel Coen  Set in 1929, a political boss and his advisor ...   
4      Akira Kurosawa  A collection of magical tales based upon the a...   
5       Ron Underwood  Val McKee and Earl Bassett are in a fight for ...   
..                ...                                                ...   
941  Michael Sarnoski  As New York City is invaded by alien creatures...   
942        Rose Glass  Reclusive gym manager Lou falls hard for Jacki...   
943   Jane Schoenbrun  Teenager Owen is just trying to make it throug...   
944     Vivek Athreya  In order to defend the innocent from a corrupt...   
945    Jithu Madhavan  Three teenagers reach Bangalore for their engi...   

                                                  Cast  \
1    ['Robert De Niro', 'Ray Liotta', 'Joe Pesci', ...   
2    ['Bruce Willis', 'Bonnie Bedelia', 'William Sa...   
3    ['Gabriel Byrne', 'Marcia Gay Harden', 'John T...   
4    ['Akira Terao', 'Mitsuko Baisho', 'Toshie Negi...   
5    ['Kevin Bacon', 'Fred Ward', 'Finn Carter', 'M...   
..                                                 ...   
941  ["Lupita Nyong'o", 'Joseph Quinn', 'Alex Wolff...   
942  ['Kristen Stewart', "Katy O'Brian", 'Ed Harris...   
943  ['Justice Smith', 'Brigette Lundy-Paine', 'Ian...   
944      ['Nani', 'S.J. Suryah', 'Priyanka Arulmohan']   
945  ['Fahadh Faasil', 'Hipzster', 'Mithun Jai Shan...   

                                                 Genre Country   Language  
1                                   ['Drama', 'Crime']     USA    English  
2                               ['Thriller', 'Action']     USA    English  
3                       ['Drama', 'Thriller', 'Crime']     USA    English  
4                                 ['Drama', 'Fantasy']     USA   Japanese  
5    ['Horror', 'Action', 'Comedy', 'Science Fiction']     USA    English  
..                                                 ...     ...        ...  
941          ['Thriller', 'Science Fiction', 'Horror']     USA    English  
942                   ['Romance', 'Thriller', 'Crime']      UK    English  
943                                ['Drama', 'Horror']     USA    English  
944                             ['Action', 'Thriller']   India     Telugu  
945                               ['Action', 'Comedy']   India  Malayalam  

[945 rows x 10 columns]

In [4]:
Movies_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 945 entries, 1 to 945
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Movie Name          945 non-null    object
 1   Movie Release Year  945 non-null    int64 
 2   Runtime             945 non-null    object
 3   Rating              945 non-null    object
 4   Director            945 non-null    object
 5   Description         945 non-null    object
 6   Cast                945 non-null    object
 7   Genre               945 non-null    object
 8   Country             945 non-null    object
 9   Language            945 non-null    object
dtypes: int64(1), object(9)
memory usage: 74.0+ KB


In [5]:
a = Movies_dataset["Runtime"].value_counts().sort_values(ascending=True)
pd.set_option('display.max_rows', None)
print(a)

206 mins             1
95 mins              1
195 mins             1
76  mins             1
180 mins             1
342 mins             1
393 mins             1
321 mins             1
202 mins             1
187 mins             1
183 mins             1
237 mins             1
83  mins             1
77  mins             1
80  mins             1
201 mins             1
82  mins             1
194 mins             1
224 mins             1
210 mins             1
216 mins             1
176 mins             2
87  mins             2
89  mins             2
182 mins             2
174 mins             2
192 mins             3
84  mins             3
No runtime found     3
185 mins             3
78  mins             3
189 mins             3
85  mins             4
88  mins             4
172 mins             4
171 mins             5
173 mins             5
161 mins             5
186 mins             5
99  mins             5
165 mins             5
144 mins             5
149 mins             5
86  mins   

In [6]:
for column in Movies_dataset.columns:
    value_counts = Movies_dataset[column].value_counts()
    print(f"Unique values and counts for coumn {column} \n {value_counts} \n")

Unique values and counts for coumn Movie Name 
 Ghajini                                                                                2
Drishyam 2                                                                             2
Drishyam                                                                               2
Vikram Vedha                                                                           2
It                                                                                     2
Original title: Qarib Qarib Singlle                                                    1
mother!                                                                                1
Spider-Man: Homecoming                                                                 1
Thor: Ragnarok                                                                         1
Dunkirk                                                                                1
GoodFellas                                                    

In [7]:
movies_with_No_Runtime = Movies_dataset[(Movies_dataset["Runtime"] == "No runtime found") | (Movies_dataset["Runtime"] == "Runtime format not recognized.")]
movies_with_No_Runtime

Movie Name  Movie Release Year           Runtime  \
138               Dil To Pagal Hai                1997  No runtime found   
593        Maheshinte Prathikaaram                2016  No runtime found   
933  Original title: Kalki 2898-AD                2024  No runtime found   

    Rating        Director                                        Description  \
138    3.5     Yash Chopra  A grand musical about three passionate and dre...   
593   4.15  Dileesh Pothan  Mahesh, a studio photographer and studio owner...   
933    3.5      Nag Ashwin  The future of those in the dystopian city of K...   

                                                  Cast  \
138  ['Shah Rukh Khan', 'Madhuri Dixit', 'Karisma K...   
593  ['Fahadh Faasil', 'Aparna Balamurali', 'Soubin...   
933    ['Prabhas', 'Amitabh Bachchan', 'Kamal Haasan']   

                                                 Genre Country   Language  
138          ['Comedy', 'Drama', 'Musical', 'Romance']   India      Hindi  
593                    ['Comedy', 'Drama', 'Thriller']   India  Malayalam  
933  ['Action Epic', 'Desert Adventure', 'Dystopian...   India     Telugu

### Mean time we found there are some movies with names like 'Original title: Kalki 2898-AD' we need to replace them to 'Kalki 2898-AD'

In [8]:
Movies_dataset['Movie Name'] = Movies_dataset['Movie Name'].str.replace("Original title: ",'')

In [9]:
movies_with_No_Runtime = Movies_dataset[(Movies_dataset["Runtime"] == "No runtime found") | (Movies_dataset["Runtime"] == "Runtime format not recognized.")]
movies_with_No_Runtime

# Problem solved

Movie Name  Movie Release Year           Runtime Rating  \
138         Dil To Pagal Hai                1997  No runtime found    3.5   
593  Maheshinte Prathikaaram                2016  No runtime found   4.15   
933            Kalki 2898-AD                2024  No runtime found    3.5   

           Director                                        Description  \
138     Yash Chopra  A grand musical about three passionate and dre...   
593  Dileesh Pothan  Mahesh, a studio photographer and studio owner...   
933      Nag Ashwin  The future of those in the dystopian city of K...   

                                                  Cast  \
138  ['Shah Rukh Khan', 'Madhuri Dixit', 'Karisma K...   
593  ['Fahadh Faasil', 'Aparna Balamurali', 'Soubin...   
933    ['Prabhas', 'Amitabh Bachchan', 'Kamal Haasan']   

                                                 Genre Country   Language  
138          ['Comedy', 'Drama', 'Musical', 'Romance']   India      Hindi  
593                    ['Comedy', 'Drama', 'Thriller']   India  Malayalam  
933  ['Action Epic', 'Desert Adventure', 'Dystopian...   India     Telugu

In [10]:
movies_with_No_name = Movies_dataset[Movies_dataset["Movie Name"] == "No title found"]
movies_with_No_name

Empty DataFrame
Columns: [Movie Name, Movie Release Year, Runtime, Rating, Director, Description, Cast, Genre, Country, Language]
Index: []

In [11]:
movies_with_No_release_year = Movies_dataset[Movies_dataset["Movie Release Year"] == "No release year found"]
movies_with_No_release_year

Empty DataFrame
Columns: [Movie Name, Movie Release Year, Runtime, Rating, Director, Description, Cast, Genre, Country, Language]
Index: []

In [12]:
movies_with_No_Director = Movies_dataset[Movies_dataset["Director"] == "No Director found"]
movies_with_No_Director

Empty DataFrame
Columns: [Movie Name, Movie Release Year, Runtime, Rating, Director, Description, Cast, Genre, Country, Language]
Index: []

In [13]:
movies_with_No_Description = Movies_dataset[Movies_dataset["Description"] == "No Description found"]
movies_with_No_Description

Empty DataFrame
Columns: [Movie Name, Movie Release Year, Runtime, Rating, Director, Description, Cast, Genre, Country, Language]
Index: []

In [14]:
movies_with_No_Cast = Movies_dataset[Movies_dataset["Cast"] == "No Cast found"]
movies_with_No_Cast

Empty DataFrame
Columns: [Movie Name, Movie Release Year, Runtime, Rating, Director, Description, Cast, Genre, Country, Language]
Index: []

In [15]:
movies_with_No_Genre = Movies_dataset[Movies_dataset["Genre"] == "No Genre found"]
movies_with_No_Genre

Empty DataFrame
Columns: [Movie Name, Movie Release Year, Runtime, Rating, Director, Description, Cast, Genre, Country, Language]
Index: []

### Renaming Genre
- Science Fiction to Sci-Fi
- Violent - Violence
- Anime - Animation

In [16]:
Movies_dataset['Genre']

1                                     ['Drama', 'Crime']
2                                 ['Thriller', 'Action']
3                         ['Drama', 'Thriller', 'Crime']
4                                   ['Drama', 'Fantasy']
5      ['Horror', 'Action', 'Comedy', 'Science Fiction']
6      ['Mystery', 'Fantasy', 'Drama', 'Romance', 'Th...
7                         ['Horror', 'Mystery', 'Drama']
8                                     ['Drama', 'Crime']
9                                        ['Documentary']
10                        ['Drama', 'Horror', 'Mystery']
11            ['Action', 'Science Fiction', 'Adventure']
12                        ['Drama', 'Thriller', 'Crime']
13                       ['Horror', 'Drama', 'Thriller']
14            ['Science Fiction', 'Comedy', 'Adventure']
15                                 ['Romance', 'Comedy']
16                       ['Drama', 'Romance', 'Fantasy']
17                                  ['Family', 'Comedy']
18                      ['Crime

In [17]:
# Finding unique Genre
Movies_dataset['Genre'].unique()

array(["['Drama', 'Crime']", "['Thriller', 'Action']",
       "['Drama', 'Thriller', 'Crime']", "['Drama', 'Fantasy']",
       "['Horror', 'Action', 'Comedy', 'Science Fiction']",
       "['Mystery', 'Fantasy', 'Drama', 'Romance', 'Thriller']",
       "['Horror', 'Mystery', 'Drama']", "['Documentary']",
       "['Drama', 'Horror', 'Mystery']",
       "['Action', 'Science Fiction', 'Adventure']",
       "['Horror', 'Drama', 'Thriller']",
       "['Science Fiction', 'Comedy', 'Adventure']",
       "['Romance', 'Comedy']", "['Drama', 'Romance', 'Fantasy']",
       "['Family', 'Comedy']", "['Crime', 'Thriller', 'Romance']",
       "['Drama']", "['Comedy', 'Drama']",
       "['Drama', 'Romance', 'Animation']",
       "['Action', 'Crime', 'Drama']", "['Drama', 'History', 'Thriller']",
       "['Crime', 'Drama']", "['Drama', 'Comedy']",
       "['Drama', 'Crime', 'Romance']", "['Comedy', 'Fantasy']",
       "['Family', 'Comedy', 'Adventure', 'Fantasy']",
       "['Thriller', 'Action', 'Scienc

In [18]:
# # Defining function to rename the genre with different name to single
# import ast

# def get_uniqe_genre(df, genre_column):
#     """
#     Standardize genre names in the dataset.
    
#     Parameters:
#     - df (pd.DataFrame): The DataFrame containing the dataset.
#     - genre_column (str): The name of the column containing genres as stringified lists.
    
#     Returns:
#     - pd.DataFrame: The DataFrame with standardized genre names.
#     """
#     genre_mapping= {
#         "Science Fiction": "Sci-Fi",
#         "Anime": "Animation",
#         "Violent": "Violence"
#     }
    
#     def rename_genres(genre_list):
#         genre_list = ast.literal_eval(genre_list)
#         standardized_list = [genre_mapping.get(genre, genre) for genre in genre_list]
#         return standardized_list
    
#     df[genre_column] = df[genre_column].apply(rename_genres)
# #     df[genre_column] = df[genre_column].apply(ast.literal_eval)
    
#     return df

In [19]:
import ast

# Define genre renaming dictionary
genre_mapping = {
    "Science Fiction": "Sci-Fi",
        "Anime": "Animation",
        "Violent": "Violence",
    "Action": "Action",
    "Comedy": "Comedy",
    # Add other genres here
}

# Function to handle genre renaming and parsing
def rename_genres(genre_list):
    # Check if the genre_list is a string (i.e., needs to be evaluated into a list)
    if isinstance(genre_list, str):
        genre_list = ast.literal_eval(genre_list)  # Convert string representation of list to list
    # Rename genres based on the mapping
    standardized_list = [genre_mapping.get(genre, genre) for genre in genre_list]
    return standardized_list

# Apply the function to the Genre column
Movies_dataset["Genre"] = Movies_dataset["Genre"].apply(rename_genres)

# Now, let's check the result
print(Movies_dataset["Genre"].head())


1                      [Drama, Crime]
2                  [Thriller, Action]
3            [Drama, Thriller, Crime]
4                    [Drama, Fantasy]
5    [Horror, Action, Comedy, Sci-Fi]
Name: Genre, dtype: object


In [21]:
import json

# Function to format the Genre column for display
def format_for_display(genre_list):
    return json.dumps(genre_list)

In [24]:
Movies_dataset["Genre"]

1                                         [Drama, Crime]
2                                     [Thriller, Action]
3                               [Drama, Thriller, Crime]
4                                       [Drama, Fantasy]
5                       [Horror, Action, Comedy, Sci-Fi]
6           [Mystery, Fantasy, Drama, Romance, Thriller]
7                               [Horror, Mystery, Drama]
8                                         [Drama, Crime]
9                                          [Documentary]
10                              [Drama, Horror, Mystery]
11                           [Action, Sci-Fi, Adventure]
12                              [Drama, Thriller, Crime]
13                             [Horror, Drama, Thriller]
14                           [Sci-Fi, Comedy, Adventure]
15                                     [Romance, Comedy]
16                             [Drama, Romance, Fantasy]
17                                      [Family, Comedy]
18                            [

In [30]:
Movies_dataset['Cast'] = Movies_dataset['Cast'].apply(lambda x: ''.join(x) if isinstance(x, list) else x)
Movies_dataset['Genre'] = Movies_dataset['Genre'].apply(lambda x: ''.join(x) if isinstance(x, list) else x)

In [31]:
Movies_dataset['Cast']

1      ['Robert De Niro', 'Ray Liotta', 'Joe Pesci', ...
2      ['Bruce Willis', 'Bonnie Bedelia', 'William Sa...
3      ['Gabriel Byrne', 'Marcia Gay Harden', 'John T...
4      ['Akira Terao', 'Mitsuko Baisho', 'Toshie Negi...
5      ['Kevin Bacon', 'Fred Ward', 'Finn Carter', 'M...
6      ['Patrick Swayze', 'Demi Moore', 'Whoopi Goldb...
7      ['Tim Robbins', 'Elizabeth Peña', 'Danny Aiell...
8      ['Hossain Sabzian', 'Monoochehr Ahankhah', 'Ma...
9      ['Pepper LaBeija', 'Octavia St. Laurent', 'Ven...
10     ['Tim Curry', 'Harry Anderson', 'Dennis Christ...
11     ['Arnold Schwarzenegger', 'Rachel Ticotin', 'S...
12     ['Al Pacino', 'Diane Keaton', 'Talia Shire', '...
13     ['James Caan', 'Kathy Bates', 'Richard Farnswo...
14     ['Michael J. Fox', 'Christopher Lloyd', 'Mary ...
15     ['Richard Gere', 'Julia Roberts', 'Jason Alexa...
16     ['Johnny Depp', 'Winona Ryder', 'Dianne Wiest'...
17     ['Macaulay Culkin', 'Joe Pesci', 'Daniel Stern...
18     ['Nicolas Cage', 'Laura 

In [32]:
Movies_dataset_combained = Movies_dataset["Movie Name"] + ' ' + Movies_dataset["Director"] + ' ' + Movies_dataset["Cast"] + ' ' + Movies_dataset["Genre"]
Movies_dataset_combained

1      GoodFellas Martin Scorsese ['Robert De Niro', ...
2      Die Hard 2 Renny Harlin ['Bruce Willis', 'Bonn...
3      Miller’s Crossing Joel Coen ['Gabriel Byrne', ...
4      Dreams Akira Kurosawa ['Akira Terao', 'Mitsuko...
5      Tremors Ron Underwood ['Kevin Bacon', 'Fred Wa...
6      Ghost Jerry Zucker ['Patrick Swayze', 'Demi Mo...
7      Jacob’s Ladder Adrian Lyne ['Tim Robbins', 'El...
8      Close-Up Abbas Kiarostami ['Hossain Sabzian', ...
9      Paris Is Burning Jennie Livingston ['Pepper La...
10     It Tommy Lee Wallace ['Tim Curry', 'Harry Ande...
11     Total Recall Paul Verhoeven ['Arnold Schwarzen...
12     The Godfather Part III Francis Ford Coppola ['...
13     Misery Rob Reiner ['James Caan', 'Kathy Bates'...
14     Back to the Future Part III Robert Zemeckis ['...
15     Pretty Woman Garry Marshall ['Richard Gere', '...
16     Edward Scissorhands Tim Burton ['Johnny Depp',...
17     Home Alone Chris Columbus ['Macaulay Culkin', ...
18     Wild at Heart David Lync

In [33]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [34]:
tfidf = TfidfVectorizer(stop_words="english")
tfidf_matrix = tfidf.fit_transform(Movies_dataset_combained)
tfidf_matrix

<945x7568 sparse matrix of type '<class 'numpy.float64'>'
	with 19146 stored elements in Compressed Sparse Row format>

In [35]:
def get_movie_recommendations(movie_title,genres, top_n = 20):
    user_input = f"{movie_title} {genres}"
    user_vector = tfidf.transform([user_input])  # Vectorize user input

    # Compute cosine similarity with all movies
    cosine_sim = cosine_similarity(user_vector, tfidf_matrix).flatten()

    # Get the top N recommendations
    top_indices = cosine_sim.argsort()[-top_n:][::-1]
    recommended_movies = Movies_dataset.iloc[top_indices]['Movie Name'].tolist()

    return recommended_movies

In [36]:
get_movie_recommendations('Home Alone', 'family, comedy')

['#Home',
 'The Addams Family',
 'Addams Family Values',
 'Home Alone',
 'Spider-Man: No Way Home',
 'Jathi Ratnalu',
 'Spider-Man: Far From Home',
 'Home Alone 2: Lost in New York',
 'Step Brothers',
 'The Hangover',
 'Easy A',
 'Superbad',
 'Dumb and Dumber',
 'Booksmart',
 'Clerks',
 'My Cousin Vinny',
 'Bottoms',
 'Zoolander',
 'Happy Gilmore',
 'Napoleon Dynamite']

In [37]:
tfidf_2 = TfidfVectorizer(min_df=3, max_features=None,
          strip_accents='unicode', analyzer='word', token_pattern= r'\w{1,}',
          ngram_range=(1, 3),
          stop_words="english")

tfidf_matrix_2 = tfidf_2.fit_transform(Movies_dataset_combained)
tfidf_matrix_2

from sklearn.metrics.pairwise import sigmoid_kernel

sigmoid = sigmoid_kernel(tfidf_matrix_2, tfidf_matrix_2)
# sigmoid[0]

# len(sigmoid)

movie_indices = pd.Series(Movies_dataset.index, index = Movies_dataset['Movie Name'])
# movie_indices

# movie_indices['Arya']

# sigmoid[293]

# list(enumerate(sigmoid[movie_indices["Arya"]]))

# sorted(list(enumerate(sigmoid[movie_indices["Arya"]])), key=lambda x: x[1], reverse=True)



def recommend_movie(movie_name):
    user_input = movie_indices[movie_name]
    
    sigmoid_score = list(enumerate(sigmoid[user_input]))
    
    sorted_sigmoid_score = sorted(sigmoid_score, key = lambda x: x[1], reverse=True)
    
    sigmoid_score_output = sorted_sigmoid_score[1:11]
    
    movies_recommended = [i[0] for i in sigmoid_score_output]
    
    return Movies_dataset['Movie Name'].iloc[movies_recommended]

recommend_movie("Pirates of the Caribbean: At World’s End")

633                          Hindi Medium
871                         Asteroid City
122                         Bottle Rocket
416                     Fantastic Mr. Fox
158                              Rushmore
547              The Grand Budapest Hotel
231                  The Royal Tenenbaums
302    The Life Aquatic with Steve Zissou
472                      Moonrise Kingdom
454                     Midnight in Paris
Name: Movie Name, dtype: object

In [38]:
from sklearn.metrics.pairwise import sigmoid_kernel
from fuzzywuzzy import process

# Recalculate the sigmoid kernel (precomputed similarity matrix)
sigmoid = sigmoid_kernel(tfidf_matrix_2, tfidf_matrix_2)

# Movie indices for easy lookup
movie_indices = pd.Series(Movies_dataset.index, index=Movies_dataset['Movie Name'])

# Function to find the closest title using fuzzy matching
def find_closest_title(input_title):
    matches = process.extract(input_title, Movies_dataset['Movie Name'], limit=1)
    if matches and matches[0][1] >= 80:  # Threshold for fuzzy matching
        return matches[0][0]
    return None

# Main recommendation function
def recommend_movie(movie_name, genres=None, use_fuzzy=False):
    recommendations = []
    fuzzy_matches = []

    # Step 1: Check if the movie exists in the dataset
    if movie_name in movie_indices:
        # Get index and calculate recommendations based on sigmoid similarity
        user_input_idx = movie_indices[movie_name]
        sigmoid_score = list(enumerate(sigmoid[user_input_idx]))
        sorted_sigmoid_score = sorted(sigmoid_score, key=lambda x: x[1], reverse=True)
        recommended_indices = [i[0] for i in sorted_sigmoid_score[1:11]]
        recommendations = Movies_dataset['Movie Name'].iloc[recommended_indices].tolist()
    elif use_fuzzy:
        # Step 2: Handle fuzzy matching if enabled and exact match not found
        closest_title = find_closest_title(movie_name)
        if closest_title:
            fuzzy_matches.append(closest_title)
            # Get index and calculate recommendations for the closest title
            user_input_idx = movie_indices[closest_title]
            sigmoid_score = list(enumerate(sigmoid[user_input_idx]))
            sorted_sigmoid_score = sorted(sigmoid_score, key=lambda x: x[1], reverse=True)
            recommended_indices = [i[0] for i in sorted_sigmoid_score[1:11]]
            recommendations = Movies_dataset['Movie Name'].iloc[recommended_indices].tolist()

    # Step 3: If no title match, recommend based on genres
    if not recommendations and genres:
        # Vectorize the genres and calculate similarity
        user_genres_vector = tfidf_2.transform([genres])
        genre_similarity = sigmoid_kernel(user_genres_vector, tfidf_matrix_2).flatten()
        genre_top_indices = genre_similarity.argsort()[-10:][::-1]
        recommendations = Movies_dataset['Movie Name'].iloc[genre_top_indices].tolist()

    return {
        'recommendations': recommendations,
        'fuzzy_matches': fuzzy_matches
    }


C:\Users\malli\Anakonda\lib\site-packages\fuzzywuzzy\fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


In [39]:
result = recommend_movie("Pirates of the Caribbean: At World’s End")
result['recommendations']


['Hindi Medium',
 'Asteroid City',
 'Bottle Rocket',
 'Fantastic Mr. Fox',
 'Rushmore',
 'The Grand Budapest Hotel',
 'The Royal Tenenbaums',
 'The Life Aquatic with Steve Zissou',
 'Moonrise Kingdom',
 'Midnight in Paris']

In [40]:
pip install fuzzywuzzy

Note: you may need to restart the kernel to use updated packages.


In [41]:
recommend_movie("Pirates of the Caribbean: At World’s End")

{'recommendations': ['Hindi Medium',
  'Asteroid City',
  'Bottle Rocket',
  'Fantastic Mr. Fox',
  'Rushmore',
  'The Grand Budapest Hotel',
  'The Royal Tenenbaums',
  'The Life Aquatic with Steve Zissou',
  'Moonrise Kingdom',
  'Midnight in Paris'],
 'fuzzy_matches': []}

In [42]:
recommend_movie("Pirates of the Caribbean: At World’s End", use_fuzzy=True)

{'recommendations': ['Hindi Medium',
  'Asteroid City',
  'Bottle Rocket',
  'Fantastic Mr. Fox',
  'Rushmore',
  'The Grand Budapest Hotel',
  'The Royal Tenenbaums',
  'The Life Aquatic with Steve Zissou',
  'Moonrise Kingdom',
  'Midnight in Paris'],
 'fuzzy_matches': []}

In [43]:
recommend_movie("Piraets of the Caribian",'action, adventure, violent', use_fuzzy=True)

{'recommendations': ['Uppi 2',
  'The Substance',
  'Avatar: The Way of Water',
  'Avatar',
  'Titanic',
  'Speed',
  'Independence Day',
  'Being John Malkovich',
  'The Lion King',
  'Twelve Monkeys'],
 'fuzzy_matches': ['Boyz n the Hood']}

In [44]:
recommend_movie("Piraets of the Caribian", use_fuzzy=True)

{'recommendations': ['Uppi 2',
  'The Substance',
  'Avatar: The Way of Water',
  'Avatar',
  'Titanic',
  'Speed',
  'Independence Day',
  'Being John Malkovich',
  'The Lion King',
  'Twelve Monkeys'],
 'fuzzy_matches': ['Boyz n the Hood']}

In [45]:
get_movie_recommendations("Pirates of the Caribbean: At World’s End", 'actoin, fantacy, adventure')

['Pirates of the Caribbean: At World’s End',
 'Pirates of the Caribbean: Dead Man’s Chest',
 'Pirates of the Caribbean: The Curse of the Black Pearl',
 'Thor: The Dark World',
 'Neon Genesis Evangelion: The End of Evangelion',
 'Jurassic World',
 'The Lost World: Jurassic Park',
 'Scott Pilgrim vs. the World',
 'The Worst Person in the World',
 'Charlie and the Chocolate Factory',
 'Saw II',
 'The 40 Year Old Virgin',
 'The Chronicles of Narnia: The Lion, the Witch and the Wardrobe',
 'Madagascar',
 'Mr. & Mrs. Smith',
 'V for Vendetta',
 'Sin City',
 'Aavesham',
 'The Descent',
 'Athadu']

In [46]:
from sklearn.metrics.pairwise import cosine_similarity

def recommend_movies(movie_title, genres=None, top_n=20):
    """
    Recommend movies based on a given title and optional genres, using cosine similarity.
    
    Parameters:
    - movie_title (str): The title of the movie to base recommendations on.
    - genres (str, optional): Genres to prioritize in recommendations.
    - top_n (int): Number of top recommendations to return.
    
    Returns:
    - list: Recommended movie titles.
    """
    # Combine the movie title and genres (if provided) into a single input string
    if genres:
        user_input = f"{movie_title} {genres}"
    else:
        user_input = movie_title
    
    # Vectorize the user input
    user_vector = tfidf.transform([user_input])  # Ensure tfidf is the trained vectorizer

    # Compute cosine similarity with all movie vectors
    cosine_sim = cosine_similarity(user_vector, tfidf_matrix).flatten()

    # If genres are provided, re-prioritize recommendations by genre match
    if genres:
        # Compute similarity of user genres with movie genres in the dataset
        genre_vector = tfidf.transform([genres])
        genre_cosine_sim = cosine_similarity(genre_vector, tfidf_matrix).flatten()
        
        # Combine scores (title match + genre match)
        combined_score = cosine_sim + 0.5 * genre_cosine_sim  # Weight genre similarity
    else:
        combined_score = cosine_sim

    # Get the top N indices with the highest combined similarity scores
    top_indices = combined_score.argsort()[-top_n:][::-1]

    # Fetch recommended movie titles
    recommended_movies = Movies_dataset.iloc[top_indices]['Movie Name'].tolist()

    return recommended_movies


In [47]:
recommend_movies("Pirates of the Caribbean: At World’s End", 'actoin, fantacy, adventure')

['Pirates of the Caribbean: At World’s End',
 'Pirates of the Caribbean: Dead Man’s Chest',
 'Pirates of the Caribbean: The Curse of the Black Pearl',
 'Thor: The Dark World',
 'Neon Genesis Evangelion: The End of Evangelion',
 'Jurassic World',
 'The Lost World: Jurassic Park',
 'Scott Pilgrim vs. the World',
 'The Worst Person in the World',
 'Charlie and the Chocolate Factory',
 'Saw II',
 'The 40 Year Old Virgin',
 'The Chronicles of Narnia: The Lion, the Witch and the Wardrobe',
 'Madagascar',
 'Mr. & Mrs. Smith',
 'V for Vendetta',
 'Sin City',
 'Aavesham',
 'The Descent',
 'Athadu']

In [48]:
recommend_movies("Pirates of the Caribbean: At World’s End")

['Pirates of the Caribbean: At World’s End',
 'Pirates of the Caribbean: Dead Man’s Chest',
 'Pirates of the Caribbean: The Curse of the Black Pearl',
 'Thor: The Dark World',
 'Neon Genesis Evangelion: The End of Evangelion',
 'Jurassic World',
 'The Lost World: Jurassic Park',
 'Scott Pilgrim vs. the World',
 'The Worst Person in the World',
 'Charlie and the Chocolate Factory',
 'Saw II',
 'The 40 Year Old Virgin',
 'The Chronicles of Narnia: The Lion, the Witch and the Wardrobe',
 'Madagascar',
 'Mr. & Mrs. Smith',
 'V for Vendetta',
 'Sin City',
 'Aavesham',
 'The Descent',
 'Athadu']

In [49]:
recommend_movies("Pirates of the Caribbean", 'fantacy, sci-fi, adventure')

['Pirates of the Caribbean: Dead Man’s Chest',
 'Pirates of the Caribbean: At World’s End',
 'Pirates of the Caribbean: The Curse of the Black Pearl',
 'Kalki 2898-AD',
 'District 9',
 'PK',
 'Minnal Murali',
 'Android Kunjappan Version 5.25',
 'Gravity',
 'Guardians of the Galaxy',
 'Dune: Part Two',
 'Independence Day',
 'Star Wars: The Last Jedi',
 'Avengers: Age of Ultron',
 'X2',
 'Thor: Ragnarok',
 'Star Wars: The Rise of Skywalker',
 'Iron Man 3',
 'Spider-Man: Homecoming',
 'The Hunger Games: Mockingjay – Part 2']

In [50]:
from fuzzywuzzy import process
from sklearn.metrics.pairwise import cosine_similarity

def recommend_movies(movie_title, genres=None, top_n=20):
    """
    Recommend movies based on a given title and optional genres, using cosine similarity.
    
    Parameters:
    - movie_title (str): The title of the movie to base recommendations on.
    - genres (str, optional): Genres to prioritize in recommendations.
    - top_n (int): Number of top recommendations to return.
    
    Returns:
    - list: Recommended movie titles.
    """
    # Check if movie title exists in the dataset
    if movie_title not in Movies_dataset['Movie Name'].values:
        print(f"Movie '{movie_title}' not found in the dataset. Trying fuzzy matching.")
        
        # Fuzzy matching to find the closest match from the movie dataset
        closest_match = process.extractOne(movie_title, Movies_dataset['Movie Name'])
        
        if closest_match:
            print(f"Did you mean '{closest_match[0]}'? Providing recommendations based on this movie.")
            movie_title = closest_match[0]  # Use the closest match found by fuzzy matching
            user_input = f"{movie_title} {genres}" if genres else movie_title  # Create user input after fuzzy match
        else:
            print("No similar movie found. Providing genre-based recommendations.")
            if not genres:
                return []  # No recommendations possible without genres
            user_input = genres  # Fallback to genre-based recommendations
    else:
        user_input = f"{movie_title} {genres}" if genres else movie_title

    # Vectorize the user input (this line will not raise an error now)
    user_vector = tfidf.transform([user_input])

    # Compute cosine similarity with all movie vectors
    cosine_sim = cosine_similarity(user_vector, tfidf_matrix).flatten()

    # If genres are provided, re-prioritize recommendations by genre match
    if genres:
        genre_vector = tfidf.transform([genres])
        genre_cosine_sim = cosine_similarity(genre_vector, tfidf_matrix).flatten()

        # Combine cosine similarity scores without any normalization
        combined_score = cosine_sim + 0.5 * genre_cosine_sim  # Weight genre similarity
    else:
        combined_score = cosine_sim

    # Get the top N indices with the highest combined similarity scores
    top_indices = combined_score.argsort()[-top_n:][::-1]

    # Exclude the input movie from recommendations
    if movie_title in Movies_dataset['Movie Name'].values:
        input_index = Movies_dataset[Movies_dataset['Movie Name'] == movie_title].index[0]
        combined_score[input_index] = -1  # Set the score to a very low value

    # Fetch recommended movie titles with additional details
    recommended_movies = Movies_dataset.iloc[top_indices][['Movie Name', 'Genre', 'Movie Release Year']].to_dict('records')
    recommended_movies_df = pd.DataFrame(recommended_movies, index=range(1, len(recommended_movies) + 1))
    return recommended_movies_df


In [51]:
recommend_movies("Pirates of the Caribbean", 'fantacy, sci-fi, adventure')

Movie 'Pirates of the Caribbean' not found in the dataset. Trying fuzzy matching.
Did you mean 'Pirates of the Caribbean: The Curse of the Black Pearl'? Providing recommendations based on this movie.


Movie Name  \
1   Pirates of the Caribbean: The Curse of the Bla...   
2                                       Kalki 2898-AD   
3          Pirates of the Caribbean: Dead Man’s Chest   
4            Pirates of the Caribbean: At World’s End   
5                                          District 9   
6                                          Iron Man 3   
7                                         Black Widow   
8                                                  PK   
9                                       Minnal Murali   
10                     Android Kunjappan Version 5.25   
11                                            Gravity   
12                            Guardians of the Galaxy   
13                                     Dune: Part Two   
14                                   Independence Day   
15                           Star Wars: The Last Jedi   
16                            Avengers: Age of Ultron   
17                                                 X2   
18                                     Thor: Ragnarok   
19                   Star Wars: The Rise of Skywalker   
20                                              Pearl   

                                                Genre  Movie Release Year  
1                              FantasyAdventureAction                2003  
2   Action EpicDesert AdventureDystopian Sci-FiSci...                2024  
3                              ActionAdventureFantasy                2006  
4                              ActionAdventureFantasy                2007  
5                                              Sci-Fi                2009  
6                               ActionAdventureSci-Fi                2013  
7                               Sci-FiActionAdventure                2021  
8                             SatireComedyDramaSci-Fi                2014  
9             SuperheroActionComedyDramaFantasySci-Fi                2021  
10           Artificial IntelligenceComedyDramaSci-Fi                2019  
11                                ThrillerDramaSci-Fi                2013  
12                              AdventureActionSci-Fi                2014  
13                                    AdventureSci-Fi                2024  
14                              AdventureActionSci-Fi                1996  
15                              AdventureActionSci-Fi                2017  
16                              ActionAdventureSci-Fi                2015  
17                              ActionAdventureSci-Fi                2003  
18                              AdventureActionSci-Fi                2017  
19                              AdventureActionSci-Fi                2019  
20                                        HorrorDrama                2022

In [52]:
recommend_movies("Arya", 'comedy, action, romance')

Movie Name  \
1                      Shiddat   
2                 True Romance   
3                     Arrambam   
4    Ponniyin Selvan: Part Two   
5                       Amaran   
6                       Arya 2   
7                    Shershaah   
8             K.G.F: Chapter 2   
9             K.G.F: Chapter 1   
10   Ponniyin Selvan: Part One   
11            The Great Father   
12                        Arya   
13               Jathi Ratnalu   
14  Salaar: Part 1 - Ceasefire   
15     Bãhubali: The Beginning   
16  Bãhubali 2: The Conclusion   
17         Sarpatta parambarai   
18                  Magadheera   
19                        Kill   
20      RRR (Rise Roar Revolt)   

                                                Genre  Movie Release Year  
1                                             Romance                2021  
2                                  RomanceCrimeAction                1993  
3                          ActionDramaMysteryThriller                2013  
4   Action EpicAdventure EpicEpicActionAdventureDrama                2023  
5                  Action EpicActionBiographyDramaWar                2024  
6                                 ActionComedyRomance                2009  
7         Action EpicDocudramaActionBiographyDramaWar                2021  
8   Action EpicEpicGangsterPeriod DramaActionCrime...                2022  
9   Action EpicEpicGangsterPeriod DramaActionCrime...                2018  
10  Action EpicAdventure EpicEpicActionAdventureDrama                2022  
11                    ActionCrimeDramaMysteryThriller                2017  
12                           ActionComedyDramaRomance                2004  
13                                             Comedy                2021  
14  Action EpicEpicOne-Person Army ActionActionCri...                2023  
15  Action EpicEpicOne-Person Army ActionPeriod Dr...                2015  
16  Action EpicEpicOne-Person Army ActionPeriod Dr...                2017  
17                             BoxingActionDramaSport                2021  
18  Action EpicFantasy EpicOne-Person Army ActionR...                2009  
19  Action EpicOne-Person Army ActionActionCrimeDr...                2023  
20  Action EpicAdventure EpicEpicPeriod DramaActio...                2022

In [53]:
from fuzzywuzzy import process
from sklearn.metrics.pairwise import cosine_similarity

def recommend_movies(movie_title, genres=None, top_n=20):
    """
    Recommend movies based on a given title and optional genres, using cosine similarity.
    
    Parameters:
    - movie_title (str): The title of the movie to base recommendations on.
    - genres (str, optional): Genres to prioritize in recommendations.
    - top_n (int): Number of top recommendations to return.
    
    Returns:
    - list: Recommended movie titles.
    """
    # Check if movie title exists in the dataset
    if movie_title not in Movies_dataset['Movie Name'].values:
        print(f"Movie '{movie_title}' not found in the dataset. Trying fuzzy matching.")
        
        # Fuzzy matching to find the closest match from the movie dataset
        closest_match = process.extractOne(movie_title, Movies_dataset['Movie Name'])
        
        if closest_match:
            print(f"Did you mean '{closest_match[0]}'? Providing recommendations based on this movie.")
            movie_title = closest_match[0]  # Use the closest match found by fuzzy matching
            user_input = f"{movie_title} {genres}" if genres else movie_title  # Create user input after fuzzy match
        else:
            print("No similar movie found. Providing genre-based recommendations.")
            if not genres:
                return []  # No recommendations possible without genres
            user_input = genres  # Fallback to genre-based recommendations
    else:
        user_input = f"{movie_title} {genres}" if genres else movie_title

    # Vectorize the user input 
    user_vector = tfidf_2.transform([user_input])

    # Compute cosine similarity with all movie vectors
    cosine_sim = cosine_similarity(user_vector, tfidf_matrix_2).flatten()

    # If genres are provided, re-prioritize recommendations by genre match
    if genres:
        genre_vector = tfidf_2.transform([genres])
        genre_cosine_sim = cosine_similarity(genre_vector, tfidf_matrix_2).flatten()

        # Combine cosine similarity scores without any normalization
        combined_score = cosine_sim + 0.5 * genre_cosine_sim  # Weight genre similarity
    else:
        combined_score = cosine_sim

    # Get the top N indices with the highest combined similarity scores
    top_indices = combined_score.argsort()[-top_n:][::-1]

    # Exclude the input movie from recommendations
    if movie_title in Movies_dataset['Movie Name'].values:
        input_index = Movies_dataset[Movies_dataset['Movie Name'] == movie_title].index[0]
        combined_score[input_index] = -1  # Set the score to a very low value

    # Fetch recommended movie titles with additional details
    recommended_movies = Movies_dataset.iloc[top_indices][['Movie Name', 'Genre', 'Movie Release Year']].to_dict('records')
    recommended_movies_df = pd.DataFrame(recommended_movies, index=range(1, len(recommended_movies) + 1))
    return recommended_movies_df

In [54]:
recommend_movies("Arya", 'comedy, action, romance')

Movie Name  \
1                                 Sarpatta parambarai   
2                                       Jathi Ratnalu   
3                                    The Great Father   
4                                           Shershaah   
5                                                Kill   
6                                              Amaran   
7                                           Booksmart   
8                                            Arrambam   
9                                       Kalki 2898-AD   
10                          Ponniyin Selvan: Part Two   
11                                   K.G.F: Chapter 1   
12                                               Arya   
13                                            Bottoms   
14                                  Napoleon Dynamite   
15                                   K.G.F: Chapter 2   
16                                             Clerks   
17  Borat: Cultural Learnings of America for Make ...   
18                                             Arya 2   
19                                    My Cousin Vinny   
20                          Ponniyin Selvan: Part One   

                                                Genre  Movie Release Year  
1                              BoxingActionDramaSport                2021  
2                                              Comedy                2021  
3                     ActionCrimeDramaMysteryThriller                2017  
4         Action EpicDocudramaActionBiographyDramaWar                2021  
5   Action EpicOne-Person Army ActionActionCrimeDr...                2023  
6                  Action EpicActionBiographyDramaWar                2024  
7                                              Comedy                2019  
8                          ActionDramaMysteryThriller                2013  
9   Action EpicDesert AdventureDystopian Sci-FiSci...                2024  
10  Action EpicAdventure EpicEpicActionAdventureDrama                2023  
11  Action EpicEpicGangsterPeriod DramaActionCrime...                2018  
12                           ActionComedyDramaRomance                2004  
13                                             Comedy                2023  
14                                             Comedy                2004  
15  Action EpicEpicGangsterPeriod DramaActionCrime...                2022  
16                                             Comedy                1994  
17                                             Comedy                2006  
18                                ActionComedyRomance                2009  
19                                             Comedy                1992  
20  Action EpicAdventure EpicEpicActionAdventureDrama                2022

In [55]:
# saving the vectors and matrix to use in web application which can help not to re-read the files again for every recommendation
import pickle

with open("tfidf_final_vectorizer.pkl", 'wb') as f:
    pickle.dump(tfidf_2,f)

with open("tfidf_final_matrix.pkl", 'wb') as f:
    pickle.dump(tfidf_matrix_2, f)

In [56]:
Movies_dataset.to_csv("Movies_dataset", index=False)
pd.read_csv("Movies_dataset")

Movie Name  Movie Release Year  \
0                                           GoodFellas                1990   
1                                           Die Hard 2                1990   
2                                    Miller’s Crossing                1990   
3                                               Dreams                1990   
4                                              Tremors                1990   
5                                                Ghost                1990   
6                                       Jacob’s Ladder                1990   
7                                             Close-Up                1990   
8                                     Paris Is Burning                1990   
9                                                   It                1990   
10                                        Total Recall                1990   
11                              The Godfather Part III                1990   
12                                              Misery                1990   
13                         Back to the Future Part III                1990   
14                                        Pretty Woman                1990   
15                                 Edward Scissorhands                1990   
16                                          Home Alone                1990   
17                                       Wild at Heart                1990   
18                        The Double Life of Véronique                1991   
19                                             My Girl                1991   
20                                      Only Yesterday                1991   
21                                          Thalapathi                1991   
22                                                 JFK                1991   
23                                         Naked Lunch                1991   
24                                         Barton Fink                1991   
25                                      Night on Earth                1991   
26                               A Brighter Summer Day                1991   
27                                   The Addams Family                1991   
28                                                Hook                1991   
29                                     Boyz n the Hood                1991   
30                          Terminator 2: Judgment Day                1991   
31                                Beauty and the Beast                1991   
32                                     Thelma & Louise                1991   
33                            The Silence of the Lambs                1991   
34                                My Own Private Idaho                1991   
35                                         Point Break                1991   
36                                           Cape Fear                1991   
37                                        Thevar Magan                1992   
38                                                Roja                1992   
39                                           Malcolm X                1992   
40                          The Muppet Christmas Carol                1992   
41                                      Basic Instinct                1992   
42                                    Scent of a Woman                1992   
43                                     My Cousin Vinny                1992   
44                                          Unforgiven                1992   
45                                            Candyman                1992   
46                                      A Few Good Men                1992   
47                                              Alien³                1992   
48                                      Reservoir Dogs                1992   
49                                             Aladdin                1992   
50                               Bram Stoker’s Dracula            